src/Bicep.Core/TypeSystem/DeployTimeConstantVisitor.cs

```diff
@@ -0,0 +1,130 @@
+        {
+            this.runtimeValueAllowed = false;
+            base.VisitResourceDeclarationSyntax(syntax);
+            this.runtimeValueAllowed = true;
+        }
+
+        public override void VisitModuleDeclarationSyntax(ModuleDeclarationSyntax syntax)
+        {
+            this.runtimeValueAllowed = false;
+            base.VisitModuleDeclarationSyntax(syntax);
+            this.runtimeValueAllowed = true;
+        }
+
+        public override void VisitObjectPropertySyntax(ObjectPropertySyntax syntax)
+        {
+            bool priorRuntimeValueAllowed = this.runtimeValueAllowed;
+            // runtimeValueAllowed should only be false if we are going through a module or resource's objectPropertySyntax
+            if (!this.runtimeValueAllowed && syntax.Key is IdentifierSyntax keyIdentifier) 
+            {
+                this.currentSymbol = keyIdentifier.IdentifierName;
+                // right now we only check for the name property (using resourceNamePropertyName but the name is same for modules)
+                if (!keyIdentifier.IdentifierName.Equals(LanguageConstants.ResourceNamePropertyName, System.StringComparison.Ordinal))
+                {
+                    this.runtimeValueAllowed = true;
+                }
+            }
+            base.VisitObjectPropertySyntax(syntax);
+            // restore the value prior to the visit
+            this.runtimeValueAllowed = priorRuntimeValueAllowed;
+        }
+
+        public override void VisitPropertyAccessSyntax(PropertyAccessSyntax syntax)
+        {
+            if (!this.runtimeValueAllowed)
+            {
+                if (syntax.BaseExpression is VariableAccessSyntax variableAccessSyntax)
+                {
+                    // validate only on resource and module symbols
+                    // TODO these switch cases are so redundant.....
+                    switch (model.GetSymbolInfo(variableAccessSyntax))
+                    {
+                        case ResourceSymbol resourceSymbol:
+                            if ((resourceSymbol.Type is ResourceType resourceType) &&
+                            (resourceType.Body is ObjectType resourceBodyObj))
+                            {
+                                var property = syntax.PropertyName.IdentifierName;
+                                if (resourceBodyObj.Properties.TryGetValue(property, out var propertyType) &&
+                                !propertyType.Flags.HasFlag(TypePropertyFlags.SkipInlining))
+                                {
+                                    AppendError(syntax);
+                                }
+                            }
+                            break;
+                        case ModuleSymbol moduleSymbol:
+                            if ((moduleSymbol.Type is ModuleType moduleType) &&
+                                (moduleType.Body is ObjectType moduleBodyObj))
```

Comment:
>ObjectType [](start = 52, length = 10)

This one is actually guaranteed to be `ObjectType` currently, but it'd be better to go through common (if feasible) logic because then we might avoid a regression if that changes.